src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -534,30 +535,48 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
 
       val lazyName  = lazyVal.name.toTermName
       val pos       = lazyVal.pos.focus
-      val flags     = (lazyVal.flags & FieldFlags | ARTIFACT | MUTABLE) & ~(IMPLICIT | STABLE) // TODO: why include MUTABLE???
-      val holderSym = owner.newValue(lazyName append nme.LAZY_LOCAL_SUFFIX_STRING, pos, flags) setInfo refTpe
+
+      // used twice: once in the same owner as the lazy val, another time inside the compute method
+      val localLazyName = lazyName append LAZY_LOCAL_SUFFIX_STRING
+
+      // The lazy holder val need not be mutable, as we write to its field.
+      // In fact, it MUST not be mutable to avoid capturing it as an ObjectRef in lambdalift
+      // Must be marked LAZY to allow forward references, as in `def test2 { println(s.length) ; lazy val s = "abc" }
+      val holderSym = owner.newValue(localLazyName, pos, LAZY | ARTIFACT) setInfo refTpe
 
       val initializedGetter = refTpe.member(nme.initialized)
       val isUnit            = refClass == LazyUnitClass
       val valueGetter       = if (isUnit) NoSymbol else refTpe.member(nme.value)
-      val getValue          = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
+      def getValue          = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
 
-      def mkChecked(res: Tree) = If(Ident(holderSym) DOT initializedGetter, getValue, res)
+      def mkChecked(alreadyComputed: Tree, compute: Tree) = If(Ident(holderSym) DOT initializedGetter, alreadyComputed, compute)
 
       val computerSym =
         owner.newMethod(lazyName append nme.LAZY_SLOW_SUFFIX, pos, ARTIFACT | PRIVATE) setInfo MethodType(Nil, lazyValType)
 
       val rhsAtComputer = rhs.changeOwner(lazyVal -> computerSym)
       val computer = mkAccessor(computerSym) {
         val setInitialized = Apply(Select(Ident(holderSym), initializedGetter.setterIn(refClass)), TRUE :: Nil)
-        val setValue =
-          if (isUnit) rhsAtComputer
-          else Apply(Select(Ident(holderSym), valueGetter.setterIn(refClass)), rhsAtComputer :: Nil)
 
-        gen.mkSynchronized(Ident(holderSym))(mkChecked(Block(List(setValue, setInitialized), getValue)))
+        // we write to a local var outside of the synchronized block to avoid boxing/unboxing (synchronized method is polymorphic)
+        val resVarSym = computerSym.newVariable(localLazyName, pos, ARTIFACT) setInfo lazyValType
+
+        if (isUnit)
+          gen.mkSynchronized(Ident(holderSym))(mkChecked(alreadyComputed = UNIT, compute = Block(List(rhsAtComputer, setInitialized), UNIT)))
+        else {
+          val alreadyComputed = Assign(Ident(resVarSym), getValue)
+          val storeComputed   = Apply(Select(Ident(holderSym), valueGetter.setterIn(refClass)), Ident(resVarSym) :: Nil)
+          val compute = Block(List(Assign(Ident(resVarSym), rhsAtComputer), storeComputed, setInitialized), Ident(resVarSym))
```

Comment:
Even when doing that, there's an unnecessary value stored from the `if`:

```
    private final /* synthetic */ int v$lzycompute$1(LazyInt v$lzy$1) {
        int v$lzy = 0;
        LazyInt lazyInt = v$lzy$1;
        synchronized (lazyInt) {
            BoxedUnit boxedUnit;
            if (v$lzy$1.initialized()) {
                v$lzy = v$lzy$1.value();
                boxedUnit = BoxedUnit.UNIT;
            } else {
                v$lzy = this.i();
                v$lzy$1.value_$eq(v$lzy);
                v$lzy$1.initialized_$eq(true);
                boxedUnit = BoxedUnit.UNIT;
            }
        }
        return v$lzy;
    }
```

Could we get rid of the `boxedUnit` local somehow? I didn't manage to build the right tree shape so far..